File: 2100-适合打劫银行的日子.cpp
#include <iostream>
using namespace std;
#include <algorithm>
#include <vector>

class Solution
{
public:
    vector<int> goodDaysToRobBank(vector<int> &security, int time)
    {
        int n = security.size();
        vector<int> res;
        // 前缀和
        // if (time == 0)
        // {
        //     for (int i = 0; i < n; i++)
        //     {
        //         res.push_back(i);
        //     }
        //     return res;
        // }
        // int left = 0, right = 0;
        // for (int i = 1; i < n - time; i++)
        // {
        //     if (security[i] <= security[i - 1])
        //     {
        //         left++;
        //     }
        //     else
        //     {
        //         left = 0;
        //     }
        //     if (security[i + time - 1] <= security[i + time])
        //     {
        //         right++;
        //     }
        //     else
        //     {
        //         right = 0;
        //     }
        //     if (left >= time && right >= time)
        //     {
        //         res.push_back(i);
        //     }
        // }
        // 动态规划
        vector<int> left(n), right(n);
        for (int i = 1; i < n; i++)
        {
            if (security[i] <= security[i - 1])
            {
                left[i] = left[i - 1] + 1;
            }
            if (security[n - i - 1] <= security[n - i])
            {
                right[n - i - 1] = right[n - i] + 1;
            }
        }
        for (int i = time; i < n - time; i++)
        {
            if (left[i] >= time && right[i] >= time)
            {
                res.push_back(i);
            }
        }
        return res;
    }
};

int main()
{
    vector<int> security = {5, 3, 3, 3, 5, 6, 2};
    int time = 2;
    Solution sol;
    vector<int> res = sol.goodDaysToRobBank(security, time);
    for (int i = 0; i < res.size(); i++)
    {
        cout << res[i] << " ";
    }
    cout << endl;
}

File: 2100-适合打劫银行的日子.py
from typing import List


class Solution:
    def goodDaysToRobBank(self, security: List[int], time: int) -> List[int]:
        # 前缀和
        # if time > 2 * len(security) + 1:
        #     return []
        # if time == 0:
        #     return list(range(len(security)))
        # res = []
        # left = right = 0
        # for i in range(1, len(security) - time):
        #     if security[i] <= security[i-1]:
        #         left += 1
        #     else:
        #         left = 0
        #     if security[i+time-1] <= security[i+time]:
        #         right += 1
        #     else:
        #         right = 0
        #     if left >= time and right >= time:
        #         res.append(i)
        # return res
        # 动态规划
        n = len(security)
        left = [0] * n
        right = [0] * n
        for i in range(1, n):
            if security[i] <= security[i-1]:
                left[i] = left[i-1] + 1
            if security[n-i-1] <= security[n-i]:
                right[n-i-1] = right[n-i] + 1
        return [i for i in range(time, n-time) if left[i] >= time and right[i] >= time]


security = [5, 3, 3, 3, 5, 6, 2]
time = 2
sol = Solution().goodDaysToRobBank(security, time)
print(sol)
